csi/moac/volume.ts

```diff
@@ -0,0 +1,877 @@
+        return;
+      }
+    }
+
+    // check that replicas are shared in the way they should be
+    let localNode: string = (this.nexus) ? this.nexus.node.name : this.publishedOn;
+    for (const nodeName in this.replicas) {
+      const replica = this.replicas[nodeName];
+      if (replica.isOffline()) {
+        continue;
+      }
+      let share;
+      const isLocal = replica.pool!.node.name === localNode;
+      if (isLocal && replica.share !== 'REPLICA_NONE') {
+        // make sure that replica that is local to the nexus is accessed locally
+        share = 'REPLICA_NONE';
+      } else if (!isLocal && replica.share === 'REPLICA_NONE') {
+        // make sure that replica that is remote to nexus can be accessed
+        share = 'REPLICA_NVMF';
+      }
+      if (share) {
+        try {
+          await replica.setShare(share);
+          delete this.nodeBlackList[nodeName];
+          // fsa will get called again because the replica was modified
+          return;
+        } catch (err) {
+          this.nodeBlackList[nodeName] = true;
+          log.error(
+            `Failed to set share protocol to ${share} for replica "${replica}": ${err}`
+          );
+        }
+      }
+    }
+
+    // If we don't have a nexus and the volume is published, then try to create one
+    if (!this.nexus) {
+      assert(this.publishedOn);
+      let nexusNode = this.registry.getNode(this.publishedOn);
+      if (nexusNode && nexusNode.isSynced()) {
+        let replicas = [];
+        for (let nodeName in this.replicas) {
+          if (!this.replicas[nodeName].isOffline() && !this.nodeBlackList[nodeName]) {
+            replicas.push(this.replicas[nodeName]);
+          }
+        }
+        if (replicas.length === 0) {
+          log.warn(`Cannot create nexus for ${this} because all replicas are bad`);
+          return;
+        }
+        try {
+          await this._createNexus(nexusNode, replicas);
+        } catch (err) {
+          log.error(`Failed to create nexus for ${this} on "${this.publishedOn}"`);
+          this._setState(VolumeState.Faulted);
+        }
+      } else {
+        log.warn(`Cannot create nexus for ${this} because "${this.publishedOn}" is down`)
+        this._setState(VolumeState.Faulted);
+      }
+      // fsa will get called again when event about created nexus arrives
+      return;
+    }
+
+    // pair nexus children with replica objects to get the full picture
+    const childReplicaPairs: {ch: Child, r: Replica | undefined}[] = this.nexus.children.map((ch) => {
+      const r = Object.values(this.replicas).find((r) => r.uri === ch.uri);
+      return {ch, r};
+    });
+    // add newly found replicas to the nexus (one by one)
+    const newReplicas = Object.values(this.replicas).filter((r) => {
+      return (!r.isOffline() &&
+        !childReplicaPairs.find((pair) => pair.r === r) &&
+        !this.nodeBlackList[r.pool!.node!.name]);
+    });
+    for (let i = 0; i < newReplicas.length; i++) {
```

Comment:
do we need to check if we've got the required number of replicas already? and if so we don't need to add these to the nexus? Also won't this cause multiple rebuilds at the same time? (which is fine but I'm not if we want that or not)

Reply:
I can think of just one case when this would be a problem (adding only as many replicas as we need) and that is when moving replicas. How that works is that we create a new replica according to new scheduling criteria and when it's added we remove the least preferred which is the old one. Now if I refused to add a child because there are already enough of them, that would break this feature.

Reply:
Ah yes that makes sense... However I still don't think we should be adding more than one at once. A while back we had a discussion about rebuild and it was pointed out that we never rebuild more than one at once so we didn't need to have more rebuild test cases for this scenario, so I'm not sure if it's a good idea to start doing that now without revisiting the tests first.

Reply:
well, if I cannot have more than one replica under rebuild then this is indeed a problem. I haven't expected that though so I will need to clarify this with you and Paul. Thanks for pointing it out.

Reply:
Paul confirmed that adding multiple children, all being in rebuild state, is ok for mayastor. Let's go with that then and we can change the code later if we see any problem with it.